pkg/controllers/streaming/processor_controller_test.go

```diff
@@ -17,44 +17,110 @@ limitations under the License.
 package streaming
 
 import (
-	"errors"
 	"testing"
+	"time"
 
 	"github.com/go-logr/logr"
-	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
-	"github.com/projectriff/system/pkg/apis"
 	"github.com/projectriff/system/pkg/apis/build/v1alpha1"
 	corev1alpha1 "github.com/projectriff/system/pkg/apis/core/v1alpha1"
 	streamingv1alpha1 "github.com/projectriff/system/pkg/apis/streaming/v1alpha1"
 	kedav1alpha1 "github.com/projectriff/system/pkg/apis/thirdparty/keda/v1alpha1"
 	rtesting "github.com/projectriff/system/pkg/controllers/testing"
-	"github.com/projectriff/system/pkg/refs"
+	"github.com/projectriff/system/pkg/controllers/testing/factories"
 	"github.com/projectriff/system/pkg/tracker"
 )
 
-const (
-	testNamespace      = "test-namespace"
-	testName           = "test-processor"
-	testProcessorImage = "test-processor-image"
-	testDefaultImage   = "test-image-from-template"
-	testFunction       = "test-function"
-	testFunctionImage  = "test-function-image"
-	testContainer      = "test-container"
-	testContainerImage = "test-container-image"
-	testErrorMessage   = "test error"
-)
+func TestReconcile(t *testing.T) {
+	scheme := runtime.NewScheme()
+	_ = clientgoscheme.AddToScheme(scheme)
+	_ = corev1alpha1.AddToScheme(scheme)
+	_ = streamingv1alpha1.AddToScheme(scheme)
+	_ = kedav1alpha1.AddToScheme(scheme)
+	_ = v1alpha1.AddToScheme(scheme)
 
-var testError = errors.New(testErrorMessage)
+	const (
+		testNamespace      = "test-namespace"
+		testName           = "test-processor"
+		testProcessorImage = "test-processor-image"
+		testDefaultImage   = "test-image-from-template"
+		testFunction       = "test-function"
+		testFunctionImage  = "test-function-image"
+		testContainer      = "test-container"
+		testContainerImage = "test-container-image"
+	)
+
+	processorGiven := factories.Processor().
+		NamespaceName(testNamespace, testName).
+		PodTemplateSpec(func(spec factories.PodTemplateSpec) {
+			spec.ContainerNamed(testContainer, func(c *corev1.Container) {
+				c.Image = testDefaultImage
+			})
+		})
+	configMapGiven := factories.ConfigMap().
+		NamespaceName(testNamespace, processorImages).
+		AddData(processorImageKey, testProcessorImage)
+	containerGiven := factories.Container().
+		NamespaceName(testNamespace, testContainer).
+		StatusLatestImage(testContainerImage)
+	deploymentGiven := factories.Deployment().
+		ObjectMeta(func(om factories.ObjectMeta) {
+			om.Namespace(testNamespace)
+			om.GenerateName("%s-processor-", testName)
```

Comment:
Adding this causes the test to fail because we aren't running with a control plane to generate the name.

Reply:
Then something else is wrong with the test. All given resources need to have a name, it can't exist on the api server without a name.

Looks like is issue is that the "given" resources are being used as the target of a create. Givens are effectively the output of a call to create, not the input.

Check out the difference between create resources and given resources here: https://github.com/projectriff/system/blob/7e4423f26389c1883b31942df141f952898741e9/pkg/controllers/knative/deployer_controller_test.go#L75-L96

Reply:
Renamed two variables. I hadn't grokked the convention.